### PR TITLE
Remove version increment from Dictionary<K,V>.Remove overloads

### PR DESCRIPTION
**Pull request by Wraith2**
Removes the version increment from Remove operations

This addresses the coreclr side of the api change [Add Dictionary<K,V>.Remove(predicate)](https://github.com/dotnet/corefx/issues/29979) with the intention of allowing removal of items from the dictionary while enumerating per direction from @jkotas . All collections tests and modified and new tests added in the related corefx PR.

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs [hunk 1]
```diff
@@ -796,7 +796,6 @@ public bool Remove(TKey key)
                         }
                         _freeList = i;
                         _freeCount++;
-                        _version++;
```
[the remark this answers is not in the document]
Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs]: As an aside I think we are inconsistent already. Looks like both TrimExcess() and EnsureCapacity() are missing version increments, even though they can invalidate an enumeration.

**Comment by stephentoub** [comment]
With the new test system @A-And implemented, ~50 tests are now failing as a result of this change and will need to be disabled in coreclr before it can be merged (they will then need to also be fixed in corefx).

File: src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs]: > Looks like both TrimExcess() and EnsureCapacity() are missing version increments, even though they can invalidate an enumeration.

Bug that we should fix?

[the remark this answers is not in the document]
Comment (Wraith2) [reply src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs]: Both could silently change enumeration results in the worst case so i'd say bug and fix.

**Comment by A-And** [comment]
To disable a test you can add it to the [exclusion list](https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/TopN.CoreFX.x64.Windows.issues.json). 

The expected format and a guide for testing can be [found in the repo](https://github.com/dotnet/coreclr/blob/master/Documentation/building/testing-with-corefx.md).

File: src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs]: https://github.com/dotnet/corefx/issues/31007

**Comment by A-And** [comment]
@Wraith2 - the parameter names are tripping up the testing framework - you'd probably want just the test names, i.e. System.Collections.Tests.Dictionary_IDictionary_NonGeneric_Tests.IEnumerable_NonGeneric_Enumerator_Reset_ModifiedAfterEnumeration_ThrowsInvalidOperationException

**Comment by danmosemsft** [comment]
@A-And the test still is failing eg https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_corefx_innerloop_prtest/164/ but seems excluded https://github.com/dotnet/coreclr/pull/18854/commits/b49ee54f43a1029ffc117de1e34f8415621e6fde#diff-2cb3a7742fa51cdfeb43b1cf4d901391R225

Can you please help us get this PR green?

**Comment by A-And** [comment]
Alright, so there seems to be two thing here -

The OSX and Ubuntu exclusions need to be added to https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/TopN.CoreFX.x64.Unix.issues.json (this just got merged yesterday). Currently they're excluded on Windows, but not on OSX or Ubuntu. That was my bad - I should have been more explicit

See my comment for the Windows failing jobs.

File: tests/CoreFX/TopN.CoreFX.x64.Windows.issues.json [hunk 2]
```diff
@@ -149,195 +149,100 @@
...
                     "reason": "Assert.All() Failure: 1 out of 4 items in the collection did not pass.\r\n"
                 },
                 {
-                    "name": "System.Collections.Tests.Dictionary_Generic_Tests_SimpleInt_int_With_Comparer_WrapStructural_SimpleInt.IEnumerable_Generic_Enumerator_Reset_ModifiedAfterEnumeration_ThrowsInvalidOperationException(count: 75)",
+                    "name": "System.Collections.Tests.Dictionary_IDictionary_NonGeneric_Tests.IEnumerable_NonGeneric_Enumerator_MoveNext_ModifiedDuringEnumeration_ThrowsInvalidOperationException",
                     "reason": "Assert.All() Failure: 1 out of 4 items in the collection did not pass.\r\n"
                 },
                 {
-                    "name": "System.Collections.Tests.Dictionary_Generic_Tests_SimpleInt_int_With_Comparer_WrapStructural_SimpleInt.IEnumerable_Generic_Enumerator_MoveNext_ModifiedAfterEnumeration_ThrowsInvalidOperationException(count: 1)",
+                    "name": " System.Collections.Tests.Dictionary_IDictionary_NonGeneric_Tests.IEnumerable_NonGeneric_Enumerator_Reset_ModifiedAfterEnumeration_ThrowsInvalidOperationException",
```
Comment (A-And) [inline tests/CoreFX/TopN.CoreFX.x64.Windows.issues.json]: This is being passed with an extra space in the beginning - the xunit runner parses it as a command arg and the test suite fails. 
I.e. if you check https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_windows_nt_corefx_innerloop_prtest/642/consoleFull you'll see:
 `error: unknown command line option: System.Collections.Tests.Dictionary_IDictionary_NonGeneric_Tests.IEnumerable_NonGeneric_Enumerator_Reset_ModifiedAfterEnumeration_ThrowsInvalidOperationException`

**Comment by danmosemsft** [comment]
We can merge now right? Still says WIP

**Comment by danmosemsft** [comment]
Could you please open a corefx issue to follow up and take advantage of this change in our corefx product code?

**Pull request merged**